antsibull/cli/doc_commands/devel.py

```diff
@@ -2,7 +2,140 @@
+
+    responses = [
+        data.download_path if isinstance(data, DownloadResults) else data for data in responses
+    ]
+    # Note: Python dicts have always had a stable order as long as you don't modify the dict.
+    # So requestors (implicitly, the keys) and responses have a matching order here.
+    return dict(zip(requestors, responses))
+
 
 def generate_docs() -> int:
-    raise NotImplementedError('Priority to implement subcommands is stable, devel, plugin, and'
-                              ' then collection commands')
+    """
+    Create documentation for the devel subcommand.
+
+    Devel documentation creates documentation for a the current development version of Ansible.
+    It uses the latest collection releases for the collections mentioned in the specified pieces
+    file to generate rst files documenting those collections.
+
+    :returns: A return code for the program.  See :func:`antsibull.cli.antsibull_docs.main` for
+        details on what each code means.
+    """
+    flog = mlog.fields(func='generate_docs')
+    flog.notice('Begin generating docs')
+
+    app_ctx = app_context.app_ctx.get()
+
+    # Parse the pieces file
+    flog.fields(deps_file=app_ctx.extra['pieces_file']).info('Parse pieces file')
+    collections = parse_pieces_file(app_ctx.extra['pieces_file'])
+    flog.debug('Finished parsing deps file')
+
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        # Retrieve ansible-base and the collections
+        flog.fields(tmp_dir=tmp_dir).info('created tmpdir')
+        collection_tarballs = asyncio_run(
+            retrieve(collections, tmp_dir,
+                     galaxy_server=app_ctx.galaxy_url,
+                     ansible_base_source=app_ctx.extra['ansible_base_source'],
+                     collection_cache=app_ctx.extra['collection_cache']))
+        # flog.fields(tarballs=collection_tarballs).debug('Download complete')
+        flog.notice('Finished retrieving tarballs')
+
+        # Get the ansible-base location
+        try:
+            ansible_base_path = collection_tarballs.pop('_ansible_base')
+        except KeyError:
+            print('ansible-base did not download successfully')
+            return 3
+        flog.fields(ansible_base_path=ansible_base_path).info('ansible-base location')
+
+        # Install the collections to a directory
+
+        # Directory that ansible needs to see
+        collection_dir = os.path.join(tmp_dir, 'installed')
+        # Directory that the collections will be untarred inside of
+        collection_install_dir = os.path.join(collection_dir, 'ansible_collections')
+        # Safe to recursively mkdir because we created the tmp_dir
+        os.makedirs(collection_install_dir, mode=0o700)
+        flog.fields(collection_install_dir=collection_install_dir).debug('collection install dir')
+
+        # Install the collections
+        asyncio_run(install_together(collection_tarballs.values(), collection_install_dir))
```

Comment:
Looking at this I wonder if you considered using anyio? It has a nice set abstractions on top of various IO loop implementations (namely asyncio, trio and curio).

Reply:
@abadger I think that's more a question for you, since you introduced asyncio :)

Reply:
I've played with trio and curio a little bit in my spare time and I like the concepts that they push better than the ones that asyncio does.  I haven't used them enough to know whether ecosystem support is needed to make good use of them, though (as in, can we use aiofiles and aiohttp directly with trio or will we need to start using a different library or will we need to write our own routines).  I stopped playing when I got to the point of working on making urwid compatible with trio (but that was a special case since urwid has its own event system which it plugs into the event loop.  Other libraries might work without needing to be ported... I don't know for sure).  I haven't looked at whether anyio implements any concepts in addition to trio; the documentation makes it looks more like you get trio-like concepts with asyncio as your backend.

I don't think switching away from asyncio right now is a good use of time, though.  The patterns of things that we're doing are pretty well established; there's most likely code somewhere else in the code base that shows you how you should be writing your code to work in parallel.  Switching to a different API will require rewriting or at least auditing those patterns to see if they work the same in the other async framework.  Perhaps in the future we could replace asyncio with something else.  It would need exploration to see just how much work it would be, though.

if anyio is truly just a way to get trio/curio-like semantics on top of asyncio, I think I'd rather port to trio directly.  But I don't know enough about anyio yet to make that call.

In any case, I think changing the async framework we're using is a separate discussion than this one.